File: bonsai_skills/src/main/java/de/unibi/citec/clf/bonsai/skills/arm/grasping/ClearPlanningScene.java
package de.unibi.citec.clf.bonsai.skills.arm.grasping;

import de.unibi.citec.clf.bonsai.actuators.PlanningSceneActuator;
import de.unibi.citec.clf.bonsai.engine.model.AbstractSkill;
import de.unibi.citec.clf.bonsai.engine.model.ExitStatus;
import de.unibi.citec.clf.bonsai.engine.model.ExitToken;
import de.unibi.citec.clf.bonsai.engine.model.config.ISkillConfigurator;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

/**
 * Clear the planning scene.
 *
 * <pre>
 *
 * Options:
 * #_KEEP_ATTACHED        [boolean] Optional (default: false)
 *      -> Specify whether collision objects attached to the robot
 *         should be kept instead of being deleted.
 *
 * Slots:
 *
 * ExitTokens:
 * success:               Planning scene successfully cleared
 * error:                 Planning scene could not be cleared
 *
 * Sensors:
 *
 * Actuators:
 *  PlanningSceneActuator
 *
 * </pre>
 *
 * @author lruegeme
 */

public class ClearPlanningScene extends AbstractSkill {
    private static final String KEY_KEEP_ATTACHED = "#_KEEP_ATTACHED";

    // used tokens
    private ExitToken tokenSuccess;
    private ExitToken tokenError;

    private PlanningSceneActuator graspAct;

    private Future<Boolean> returnFuture;

    private boolean keepAttached;

    @Override
    public void configure(ISkillConfigurator configurator) {
        tokenSuccess = configurator.requestExitToken(ExitStatus.SUCCESS());
        tokenError = configurator.requestExitToken(ExitStatus.ERROR());

        keepAttached = configurator.requestOptionalBool(KEY_KEEP_ATTACHED, false);

        graspAct = configurator.getActuator("PlanningSceneActuator", PlanningSceneActuator.class);
    }

    @Override
    public boolean init() {
        returnFuture = graspAct.clearScene(keepAttached);
        logger.debug("clearing planning scene...");

        return true;
    }

    @Override
    public ExitToken execute() {
        if (!returnFuture.isDone()) {
            return ExitToken.loop();
        }
        try {
            if (returnFuture.get()) {
                return tokenSuccess;
            } else {
                return tokenError;
            }
        } catch (InterruptedException | ExecutionException e) {
            return tokenError;
        }
    }

    @Override
    public ExitToken end(ExitToken curToken) {
        if(returnFuture != null && !returnFuture.isDone()) {
            returnFuture.cancel(true);
        }
        return curToken;
    }
}

File: bonsai_skills/src/main/java/de/unibi/citec/clf/bonsai/skills/ecwm/grasping/ClearPlanningScene.kt
package de.unibi.citec.clf.bonsai.skills.ecwm.grasping

import de.unibi.citec.clf.bonsai.engine.model.AbstractSkill
import de.unibi.citec.clf.bonsai.engine.model.ExitStatus
import de.unibi.citec.clf.bonsai.engine.model.ExitToken
import de.unibi.citec.clf.bonsai.engine.model.config.ISkillConfigurator
import de.unibi.citec.clf.bonsai.actuators.ECWMActuator
import de.unibi.citec.clf.bonsai.actuators.ECWMGrasping
import de.unibi.citec.clf.bonsai.core.`object`.MemorySlotReader
import de.unibi.citec.clf.btl.data.ecwm.EntityList

import java.util.concurrent.Future

class ClearPlanningScene : AbstractSkill() {

    private val KEY_CLEARATTACHED= "clear_attached"

    private var fur: Future<Boolean>? = null
    private var ecwm: ECWMGrasping? = null
    private var tokenSuccess: ExitToken? = null

    private var clear_attached = false

    override fun configure(configurator: ISkillConfigurator) {
        tokenSuccess = configurator.requestExitToken(ExitStatus.SUCCESS())

        ecwm = configurator.getActuator("ECWMGrasping", ECWMGrasping::class.java)

        clear_attached = configurator.requestOptionalBool(KEY_CLEARATTACHED, clear_attached)

    }

    override fun init(): Boolean {
        fur = ecwm?.setupPlanningScene(listOf(),true,clear_attached)

        return fur != null
    }

    override fun execute(): ExitToken {
        while (!fur!!.isDone) {
            return ExitToken.loop()
        }
        return if (fur!!.get()) tokenSuccess!! else ExitToken.fatal()
    }

    override fun end(curToken: ExitToken): ExitToken {
        return curToken
    }
}
